wallet: crash fix, handle non-writable db directories

furszy opened the pull request:
Make wallet creation and load fail with a clear error when the db directory isn’t writable.

#### 1) For Wallet Creation

Before: creating a wallet would return a generic error:
"SQLiteDatabase: Failed to open database: unable to open database file"

After: creating a wallet returns:
"SQLiteDatabase: Failed to open database in directory <dir_path>: directory is not writable"

#### 2) For Wallet Loading

We currently allow loading wallets located on non-writable directories. This is problematic
because the node crashes on any subsequent write; generating a block is enough to trigger it.
Can be verified just by running the following test on master: https://github.com/furszy/bitcoin-core/commit/85fa4e2910f0a8f43d4ee5a1322cbb69656d85e9 

Also, to check directory writability, this creates a tmp file rather than relying on the
`permissions()` functions, since perms bits alone may not reliably reflect actual writability
in some systems.

Testing Note:
Pushed the tests in separate commits so they can be cherry-picked on master for comparison.

furszy committed:
wallet: handle non-writable db directories

1) For wallet load, this fixes a crash.

We currently allow loading wallets located on non-writable directories.
This is problematic because the node crashes on any subsequent write.
E.g. generating a block is enough to trigger it.

2) For wallet creation, this improves the returned error msg.

Before: creating a wallet would return a generic error:
"SQLiteDatabase: Failed to open database: unable to open database file"

After: creating a wallet returns:
"SQLiteDatabase: Failed to open database in directory <dir_path>: directory
is not writable"

furszy committed:
test: add coverage for wallet creation in non-writable directory

rkrux commented on test/functional/wallet_createwallet.py:
```diff
@@ -25,9 +28,26 @@ def set_test_params(self):
+        except PermissionError:
+            pass
+        assert_raises_rpc_error(-4, f"SQLiteDatabase: Failed to open database in directory '{str(dir_path)}': directory is not writable", node.createwallet, wallet_name=wallet_name, descriptors=True)
+        dir_path.chmod(dir_path.stat().st_mode | stat.S_IWRITE)
```
By debugging the permissions at 3 different stages here, it appears that the original permissions of the directory are not set at the end of this test.

```diff
diff --git a/test/functional/wallet_createwallet.py b/test/functional/wallet_createwallet.py
index 3bd8e6a310..29a2db3dab 100755
--- a/test/functional/wallet_createwallet.py
+++ b/test/functional/wallet_createwallet.py
@@ -33,7 +33,9 @@ class CreateWalletTest(BitcoinTestFramework):
         wallet_name = "bad_permissions"
         dir_path = node.wallets_path / wallet_name
         dir_path.mkdir(parents=True)
+        print(dir_path.stat().st_mode)
         os.chmod(dir_path, stat.S_IREAD | stat.S_IEXEC)
+        print(dir_path.stat().st_mode)
         try:
             (dir_path / f".tmp_{random.randrange(1 << 32)}").touch() # Verify the directory is actually non-writable
             self.log.warn("Skipping non-writable directory test: unable to enforce read-only permissions")
@@ -42,12 +44,13 @@ class CreateWalletTest(BitcoinTestFramework):
             pass
         assert_raises_rpc_error(-4, f"SQLiteDatabase: Failed to open database in directory '{str(dir_path)}': directory is not writable", node.createwallet, wallet_name=wallet_name, descriptors=True)
         dir_path.chmod(dir_path.stat().st_mode | stat.S_IWRITE)
+        print(dir_path.stat().st_mode)
```

```zsh
➜  bitcoin git:(2025_wallet_check_db_permissions) ✗ ./build/test/functional/wallet_createwallet.py            
2025-12-30T09:01:50.770611Z TestFramework (INFO): PRNG seed is: 3561247270765759257
2025-12-30T09:01:50.771467Z TestFramework (INFO): Initializing test directory /var/folders/6v/mpspb4bx4zzf3xr2z96hgq9h0000gn/T/bitcoin_func_test_5m4h9kjq
2025-12-30T09:01:51.370097Z TestFramework (INFO): Test wallet creation failure due to non-writable directory
16877
16704
16832
2025-12-30T09:01:51.424150Z TestFramework (INFO): Stopping nodes
2025-12-30T09:01:51.689178Z TestFramework (INFO): Cleaning up /var/folders/6v/mpspb4bx4zzf3xr2z96hgq9h0000gn/T/bitcoin_func_test_5m4h9kjq on exit
2025-12-30T09:01:51.689295Z TestFramework (INFO): Tests successful
```

I believe 16877 (0o40755) should be set at the end of this test

rkrux commented on test/functional/wallet_createwallet.py:
```diff
@@ -25,9 +28,26 @@ def set_test_params(self):
+        os.chmod(dir_path, stat.S_IREAD | stat.S_IEXEC)
+        try:
+            (dir_path / f".tmp_{random.randrange(1 << 32)}").touch() # Verify the directory is actually non-writable
+            self.log.warn("Skipping non-writable directory test: unable to enforce read-only permissions")
```
Based on the logs, it appears that all the 9 failing jobs skipped the test. At the moment, I don't know why the `touch` function didn't throw here even after non-writable permissions given to the directory. The jobs failed because the clean up was not done after the test because of access denied errors. Restoring the original permissions of the directory in a `finally` block should help here.

```zsh
stderr:
D:\a\bitcoin\bitcoin\build\test\functional\wallet_createwallet.py:39: DeprecationWarning: The 'warn' method is deprecated, use 'warning' instead

  self.log.warn("Skipping non-writable directory test: unable to enforce read-only permissions")

Traceback (most recent call last):

  File "D:\a\bitcoin\bitcoin\build\test\functional\wallet_createwallet.py", line 202, in <module>

    CreateWalletTest(__file__).main()

    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^

  File "D:\a\bitcoin\bitcoin\build\test\functional\test_framework\test_framework.py", line 154, in main

    exit_code = self.shutdown()

  File "D:\a\bitcoin\bitcoin\build\test\functional\test_framework\test_framework.py", line 332, in shutdown

    shutil.rmtree(self.options.tmpdir)

    ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^

  File "C:\hostedtoolcache\windows\Python\3.14.2\x64\Lib\shutil.py", line 852, in rmtree

    _rmtree_impl(path, dir_fd, onexc)

    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^

  File "C:\hostedtoolcache\windows\Python\3.14.2\x64\Lib\shutil.py", line 697, in _rmtree_unsafe

    onexc(os.rmdir, fullname, err)

    ~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^

  File "C:\hostedtoolcache\windows\Python\3.14.2\x64\Lib\shutil.py", line 693, in _rmtree_unsafe

    os.rmdir(fullname)

    ~~~~~~~~^^^^^^^^^^

PermissionError: [WinError 5] Access is denied: 'D:\\a\\_temp\\test_runner_₿_🏃_20251230_063644\\wallet_createwallet_186\\node0\\regtest\\wallets\\bad_permissions'
```

rkrux reviewed (commented):
Combining the individual suggestions, the full diff:

```diff
diff --git a/test/functional/wallet_createwallet.py b/test/functional/wallet_createwallet.py
index 3bd8e6a310..f0b34d685d 100755
--- a/test/functional/wallet_createwallet.py
+++ b/test/functional/wallet_createwallet.py
@@ -33,21 +33,23 @@ class CreateWalletTest(BitcoinTestFramework):
         wallet_name = "bad_permissions"
         dir_path = node.wallets_path / wallet_name
         dir_path.mkdir(parents=True)
-        os.chmod(dir_path, stat.S_IREAD | stat.S_IEXEC)
+        original_dir_perms = dir_path.stat().st_mode
+        os.chmod(dir_path, original_dir_perms & ~(stat.S_IWUSR | stat.S_IWGRP | stat.S_IWOTH))
         try:
-            (dir_path / f".tmp_{random.randrange(1 << 32)}").touch() # Verify the directory is actually non-writable
-            self.log.warn("Skipping non-writable directory test: unable to enforce read-only permissions")
-            return
+            if (dir_path / f".tmp_{random.randrange(1 << 32)}").touch(): # Verify the directory is actually non-writable
+                self.log.warning("Skipping non-writable directory test: unable to enforce read-only permissions")
+                return
+            assert_raises_rpc_error(-4, f"SQLiteDatabase: Failed to open database in directory '{str(dir_path)}': directory is not writable", node.createwallet, wallet_name=wallet_name, descriptors=True)
         except PermissionError:
             pass
-        assert_raises_rpc_error(-4, f"SQLiteDatabase: Failed to open database in directory '{str(dir_path)}': directory is not writable", node.createwallet, wallet_name=wallet_name, descriptors=True)
-        dir_path.chmod(dir_path.stat().st_mode | stat.S_IWRITE)
+        finally:
+            dir_path.chmod(original_dir_perms)
 
```

furszy commented on test/functional/wallet_createwallet.py:
```diff
@@ -25,9 +28,26 @@ def set_test_params(self):
+        os.chmod(dir_path, stat.S_IREAD | stat.S_IEXEC)
+        try:
+            (dir_path / f".tmp_{random.randrange(1 << 32)}").touch() # Verify the directory is actually non-writable
+            self.log.warn("Skipping non-writable directory test: unable to enforce read-only permissions")
```
Done. Thanks.

furszy commented on test/functional/wallet_createwallet.py:
```diff
@@ -25,9 +28,26 @@ def set_test_params(self):
+        wallet_name = "bad_permissions"
+        dir_path = node.wallets_path / wallet_name
+        dir_path.mkdir(parents=True)
+        os.chmod(dir_path, stat.S_IREAD | stat.S_IEXEC)
```
Done as suggested. Thanks.

furszy commented:
> the changes necessary to do that here seem a bit excessive for the pay-off? It doesn't seem like this change would have made resolving #34163 trivial?

It turned out not to help with #34163, but it is a general improvement to reporting read-only permission issues. Which we currently don't do properly. Can take #34163 as the starting point that led me to uncover other issues.

Something I didn’t mention in the PR description is that this also fixes opening a non-writable wallet directory, which is actually more important than issues during creation. We currently open the wallet even when the directory is read-only, and then crash on any subsequent write because we cannot write to the `-journal` file. I should add a simple test case demonstrating this bad behavior and append it to the PR description..

Also, do you really think that labeling these changes as "excessive" is correct?
We are talking about an 8-line function that improves the user experience at almost no cost and could be useful in other contexts as well. Even setting aside the post-opening crash and only focusing on the creation issue, I wouldn’t label this as an excessive change.

> Didn't test, but would `sqlite3_errmsg(m_db)` offer a more simple alternative? I think I'm slightly Approach NACK atm, unless I'm wrong on the usefulness of the change.

`sqlite3_errmsg(m_db)` returns the same message as `sqlite3_errstr(ret)`. Both return the generic and not particularly useful "unable to open database file".

furszy committed:
test: add coverage for loading a wallet in a non-writable directory

Previously, wallets in non-writable directories were loaded,
leading to crashes on any subsequent write.

stickies-v commented:
> Also, do you really think that labeling these changes as "excessive" is correct?

It's not a big diff, but adding helper functions to improve diagnostic messages for a rare issue that is reasonably straightforward for the user to diagnose felt like perhaps not something we should have in our codebase, hence why I used "excessive" (in a relative way).

> The node crashes after loading a wallet in a non-writable directory; generating a block is enough to trigger it.

This seems like a much better rationale to me for this PR. Concept ACK. Approach-wise, I'm confused why the `sqlite3_db_readonly` call we already have doesn't do what we'd expect it to do, but I'm not sure I'll be diving into this further, just wanted to leave early concept feedback.

bensig commented:
ACK 2ef6adde86f9a248463b8b628f559012be103382 

Tests passed

maflcko commented on src/util/fs_helpers.cpp:
```diff
@@ -304,6 +305,21 @@ std::optional<fs::perms> InterpretPermString(const std::string& s)
+    if (!fs::is_directory(dir_path)) throw std::runtime_error(strprintf("Path %s is not a directory", fs::PathToString(dir_path)));
+    FastRandomContext rng;
+    const auto tmp = dir_path / fs::PathFromString(strprintf(".tmp_%d", rng.rand64()));
+    if (auto created{fsbridge::fopen(tmp, "a")}) {
```
would need to be 

```cpp
#ifdef __MINGW64__
            "w" // Temporary workaround for https://github.com/bitcoin/bitcoin/issues/30210
#else
            "wx"
#endif

rkrux reviewed (approved):
lgtm ACK 271846280db49654680a5d7e7948f37875e4738e

```
git range-diff c22de0d82ed87846a40629e40746b696621da3c7...271846280db49654680a5d7e7948f37875e4738e
```
Sorry for the delay, this PR had fallen off my radar somehow.

furszy commented:
Small push to fix the IWYU CI job. Ready to go.

rkrux commented:
re-ACK 5272012

```
git range-diff 2718462...5272012
```

furszy commented:
rebased, conflicts solved. Ready to go.

rkrux commented:
re-ACK 08925d5ee75c15712f523bacbdcfd44257f9c8ac

```zsh
git range-diff 5272012...08925d5
```

seduless reviewed (commented):
Tested ACK 08925d5ee75c15712f523bacbdcfd44257f9c8ac

nit: Looks like the `finally` block from the earlier discussion was dropped during a rebase, so the cleanup is unprotected again in both `wallet_startup.py` and `wallet_createwallet.py`.

achow101 commented:
ACK 08925d5ee75c15712f523bacbdcfd44257f9c8ac
